Implement air quality alerts in styled forecasts view

patrickjfl opened the pull request:
## Changes in this PR

This PR implements air quality alerts within the air pollution tabs, as well as rendering guidance when there is an alert in place. The styles are updated for these components, as is the implementation of turbo stream, allowing us to update multiple HTML elements with a single request.

## Screenshots of UI changes

### Before

![Screenshot 2024-11-01 at 11 54 00](https://github.com/user-attachments/assets/adc5cee1-9fa7-4dae-9e15-8ffbe677a20c)

### After

![Screenshot 2024-11-01 at 11 53 21](https://github.com/user-attachments/assets/36bbac84-8c26-46ac-af1e-5db35387c030)

## Next steps

- [ ] Run any necessary
      [accessibility testing](https://playbook.dxw.com/guides/web-accessibility.html)
      on this feature.

patrickjfl committed:
Add air quality alert styling colours to Tailwind

As these class names will be dynamically generated in the template, we need to explicitly include all of them in our Tailwind config file ‘safelist’, or else they will not be included in the build: https://tailwindcss.com/docs/content-configuration#safelisting-classes

patrickjfl committed:
Add integration tests for air quality alerts

patrickjfl committed:
Refactor turbo stream implementation for updating forecasts

This approach, of returning via a turbo stream template, gives us more flexibility to apply changes to multple distinct elements within the page. In our case, we want to re render the pollen, UV and temperature predictions, but we’ll also want to re render the  alert guidance component in the next commit.

patrickjfl committed:
Add exclamation triangle icon

We will use this in our air pollution tabs when there is an alert.

patrickjfl committed:
Update JS active/inactive class toggle

We now need to toggle an inactive class as well as an active class, because of the interaction of the active class with the daqi-level classes. In cases of alerts, we want to override the active class with the relevant daqi-level class, but where that tab is not active, we want to override that again with our inactive styling.

patrickjfl committed:
Replace day_tab partial with new view component

patrickjfl committed:
Add styles for incorporating alerts into air pollution tabs

The rules that govern which styles should apply are quite complex, so I have ended up with a slightly verbose implementation that hopefully is at least fairly clear.

The order in which the classes are listed is also important - the ‘inactive’ class needs to be below ‘active’ in order to override it.

There are two versions of each of the DAQI alert classes. This is unfortunate but allows us to simplify the Javascript code that toggles the selected classes.

patrickjfl committed:
Fix Cuprite error

The integration tests in CI were complaining about overlapping elements on a button because of he icon on the new air pollution tab. It doesn’t matter that it overlaps, so I’ve introduced this alternative syntax to fix the issue.

edavey commented on tailwind.config.js:
```diff
@@ -5,5 +5,20 @@ module.exports = {
+  theme: {
+    extend: {
+      colors: {
+        'moderate-alert-secondary': '#FBDA3033',
```
what does "secondary" mean in this context?

patrickjfl replied:
mmmm, good question. As far as I can see, we're using the full colour range for the air pollution tab itself when there's an alert, but only using two colours for the guidance panel. So I've listed those as 'secondary'. But maybe they would make more sense as `moderate-alert-guidance-panel` or something. I'll change them 👍

edavey commented on tailwind.config.js:
```diff
@@ -5,5 +5,20 @@ module.exports = {
+    }
+  }
+  },
+  safelist: [
```
can you explain this "safelist"?

patrickjfl replied:
is it not explained by the commit message? Happy to add more if it's unclear, but I'm not sure as it is what to add.

edavey replied:
perhaps it will be helpful to include a link to the Tailwind docs https://tailwindcss.com/docs/content-configuration#safelisting-classes so that readers can understand that this "safelist" is a last resort for scenarios where tailwind can't discover classes through the `content` configuration? (I had to google to understand...)

patrickjfl replied:
thanks, yes good point. have added this 👍

edavey commented on app/models/air_pollution_prediction.rb:
```diff
@@ -27,6 +27,23 @@ def initialize(
     @label = label
   end
 
+  TAG_COLOURS = {
+    1 => "bg-lime-400",
+    2 => "bg-green-400",
+    3 => "bg-lime-600",
+    4 => "bg-yellow-300",
+    5 => "bg-amber-200",
+    6 => "bg-yellow-500",
+    7 => "bg-orange-500",
+    8 => "bg-red-500",
+    9 => "bg-red-800",
+    10 => "bg-stone-700"
+  }
+
+  def daqi_indicator_colour_class
+    TAG_COLOURS.fetch(value)
+  end
+
```
Is there a reason why you didn't want or weren't able to put this presentation code in a component e.g `DayTabComponent` or a `DaqiMarkerComponent`? To me that seems like the right location: the component should be responsible for presenting the data which is found in `AirPollutionPrediction` and that the model shouldn't concern itself with presentation. For example there might be multiple places / UIs / themes / platforms where this model is presented...

patrickjfl replied:
ah of course, i forgot about view components! I was wrestling with how awkward it would be to have that logic in a template but of course a view component is the obvious solution. thanks 💡

edavey commented on spec/factories/air_pollutions_predictions.rb:
```diff
@@ -53,6 +53,96 @@
       ozone { 10 }
     end
 
+    trait 1 do
```
nice use of FactoryBot traits!

edavey commented on spec/models/air_pollution_prediction_spec.rb:
```diff
@@ -0,0 +1,103 @@
+    context "when the #value is 1" do
+      let(:prediction) do
+        FactoryBot.build(:air_pollution_prediction, 1)
+      end
```
As I said above, I think this colour behaviour ought to be added by a view component, and so our "class under test" would be the component e.g.:
```rb
let(:prediction_component) do
  render_inline(DaqiMarkerComponent.new(
    prediction: FactoryBot.build(:air_pollution_prediction, 1)
  ))
end
```

edavey commented on app/javascript/controllers/tab_controller.js:
```diff
@@ -2,19 +2,21 @@ import { Controller } from "@hotwired/stimulus";
+    document.querySelectorAll(".tabs .tab").forEach((el) => {
+      el.classList.remove(this.activeClass);
+      el.classList.add(this.inactiveClass);
+    });
```
I think we should document this in our integration test? I think that should be straightforward to do? Adding some ".inactive" assertions to:

```rb
  def then_i_see_that_the_tomorrow_tab_is_active
    expect(page).to have_css(".tab.tomorrow.active")

    expect(page).not_to have_css(".tab.today.active")
    expect(page).not_to have_css(".tab.day_after_tomorrow.active")
  end

  def and_i_see_that_the_today_tab_is_active
    expect(page).to have_css(".tab.today.active")

    expect(page).not_to have_css(".tab.tomorrow.active")
    expect(page).not_to have_css(".tab.day_after_tomorrow.active")
  end

  def then_i_see_that_the_day_after_tomorrow_tab_is_active
    expect(page).to have_css(".tab.day_after_tomorrow.active")

    expect(page).not_to have_css(".tab.today.active")
    expect(page).not_to have_css(".tab.tomorrow.active")
  end
```

patrickjfl replied:
good shout

edavey commented on app/views/styled_forecasts/_day_tab.html.erb:
```diff
@@ -16,8 +16,11 @@
+    <div class="size-9 rounded-full <%= forecast.air_pollution.daqi_indicator_colour_class %>"></div>
+    <% if forecast.air_pollution.value > 6 %>
+      <%= render partial: "shared/icons/exclamation_triangle" %>
+    <% end %>
```
What do you think about using a `DaqiIndicatorComponent` which is initialised with an `AirPollutionPrediction`?

Perhaps we then ought to go further an extract out the "day tab" markup to a `DayTabComponent`?

patrickjfl replied:
I've now extracted the entire partial into a DayTabComponent. How much value do you think there is in creating an additional sub component for the daqi indicator?

edavey reviewed (commented):
Really nice work, very clearly structured. 

However, I do think it's wrong for the `#daqi_indicator_colour_class` to reside in `AirPollutionPrediction` --  that it in fact belongs in a view component. Perhaps the right view component would be `DaqiMarkerComponent`?

patrickjfl committed:
Create new DayTab view component

This component will allow us to store the logic required to calculate the styling required, which vary according to each DAQI level.

Add background colour class util to air pollution prediction

We will use this to generate the colour used in the forecast tabs.

patrickjfl committed:
Create partial for rendering alert guidance

This will be triggered by the turbo stream in the previous commit, and will be rendered when there is an air quality alert.

I have also removed the bold tags from the guidance text as they are not part of the designs.

edavey commented on app/components/day_tab_component.html.erb:
```diff
@@ -0,0 +1,33 @@
+<%= tag.div class: "tab py-4 flex-1 #{@day} mx-2 px-1 text-center #{@day == :today ? "active daqi-level-#{@forecast.air_pollution.value}-today" : 'inactive after-today'}",
```
It might be nice to extract this ternary condition 

`@day == :today ? "active daqi-level-#{@forecast.air_pollution.value}-today" : 'inactive after-today'` 

to the component e.g. as `DayTabComponent#classes`

and then layout the rules clearly in the unit test.

I looks to me like this expression is doing more than one thing based on whether the active tab is "today":

- setting active or inactive
- setting the daqi score
- indicating whether the tab is for today or not

Perhaps this should be done in separate, simpler, methods? That might drive out a simpler CSS also...

patrickjfl replied:
good shout, done ✅

edavey reviewed (approved):
Looking good! 

Now that the styling of the day tabs is becoming more complex it seems like a good opportunity to move that logic into the component -- and see if perhaps there are actually several class-related methods needed, rather than one very complicated one! This in turn may lead to simpler CSS rules?

patrickjfl committed:
Update alert stylings for air pollution tabs

If today has an alert, we always want to display the alert stylings. For tomorrow and the day after tomorrow, if they have alerts, we only want to display the alert stylings when the tab is selected.

I have opted for a slightly verbose Javascript implementation that extends the Stimulus code that was already in place, rather than bringing the tabs into the Turbo stream that is already in place for this action. However, if any part of this design becomes any more complex, this would probably be a good candidate to refactor along those lines.

patrickjfl commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Implement air quality alerts in styled forecasts view](https://trello.com/c/WUEWEMzJ/94-implement-air-quality-alerts-in-styled-forecasts-view)

edavey commented on app/javascript/controllers/tab_controller.js:
```diff
@@ -3,14 +3,21 @@ import { Controller } from "@hotwired/stimulus";
-    this.dayTarget.classList.toggle(this.activeClass);
-    this.dayTarget.classList.toggle(this.inactiveClass);
+    this.removeDaqiAlertClasses();
+    if (
+      this.dayTarget.innerText !== "Today" &&
+      this.daqiValueTarget.innerText[6] > 3
+    ) {
+      this.addDaqiAlertClass();
+    }
```
this might perhaps be clearer if you extracted to a function with an intention-revealing name e.g. `addAlertClassIfTodayHasAirQualityAlert()`?

patrickjfl replied:
yes, although it's more like `addAlertClassIfNotTodayAndHasAirQualityAlert` 😄

patrickjfl replied:
maybe that is ok though, even if it is long

edavey commented on app/javascript/controllers/tab_controller.js:
```diff
@@ -19,4 +26,28 @@ export default class extends Controller {
+    );
+  }
+
+  removeDaqiAlertClasses() {
+    document.querySelectorAll(".tabs .tab").forEach((el) => {
+      const daqiAlertClassRegex = new RegExp(
+        "daqi-alert-after-today-selected-level-(\\d{1,2})"
+      );
+
+      const fullClassName = Array.from(el.classList).find((className) =>
+        className.match(daqiAlertClassRegex)
+      );
+
+      if (fullClassName) {
+        el.classList.remove(fullClassName);
+      }
+    });
```
clever! bravo 👏

edavey reviewed (approved):
Really interesting to see how you've extended the Stimulus controller to handle this complexity. As you note in your commit message, you may find it easier to re-render from the server when you next come to add more complexity to this area!

patrickjfl committed:
Update integration tests to check for alert classes

We need to update our integration tests to allow us to explicitly set DAQI values, so that we can check that the correct class for that value has been set on the active tab where there is an alert.
